[Dev][Build] VS 2026 Support

snickler opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR updates the PowerToys solution to support **Visual Studio 2026 (PlatformToolset v145)**. It centralizes the build configuration, updates the C++ language standards, and fixes an issue with a MouseJump unit test that appears while using the VS 2026 supported build agent.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [ ] Closes: #xxx
- [x] **Communication:** I've discussed this with core contributors already. If the work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end-user-facing strings can be localized
- [x] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed, or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

**Build System & Configuration:**
- Updated `Cpp.Build.props` to use `v145` (VS 2026) as the default `PlatformToolset`, with fall back to `v143` for VS 2022.
- Configured C++ Language Standard:
  - `stdcpplatest` for production projects.
- Removed explicit `<PlatformToolset>` definitions from individual project files (approx. 37 modules) to inherit correctly from the central `Cpp.Build.props`.

**Code Refactoring & Fixes:**
- Updated `DrawingHelperTests.cs` in MouseJump Unit Test to ease the pixel difference tolerance. This became an issue after switching to the new VS2026 build agent.
<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

- Validated successful compilation of the entire solution. Similar updates have been made to the .NET 10 branch, but these are much cleaner and will be merged into that branch once fully confirmed working.

snickler committed:
[build] Update PlatformToolset to v145 and configure C++ standard/coroutines

snickler committed:
[logging] Removed PlatformToolset from project

snickler committed:
[installer] Removed PlatformToolset from project

snickler committed:
[PackageIdentity] Removed PlatformToolset from project

snickler committed:
[Update] Removed PlatformToolset from project

snickler committed:
[common] Removed PlatformToolset from project

snickler committed:
[ActionRunner] Removed PlatformToolset from project

snickler committed:
[EnvironmentVariables] Removed PlatformToolset from project

snickler committed:
[CropAndLock] Removed PlatformToolset from project

snickler committed:
[Hosts] Removed PlatformToolset from project

snickler committed:
[FileLocksmith] Removed PlatformToolset from project

snickler committed:
[AdvancedPaste] Removed PlatformToolset from project

snickler committed:
[MouseWithoutBorders] Removed PlatformToolset from project

snickler committed:
[NewPlus] Removed PlatformToolset from project

snickler committed:
[LightSwitch] Removed PlatformToolset from project

snickler committed:
[MeasureTool] Removed PlatformToolset from project

snickler committed:
[PowerOCR] Removed PlatformToolset from project

snickler committed:
[MouseUtils] Removed PlatformToolset from project

snickler committed:
[Workspaces] Removed PlatformToolset from project

snickler committed:
[cmdNotFound] Removed PlatformToolset from project

snickler committed:
[awake] Removed PlatformToolset from project

snickler committed:
[alwaysontop] Removed PlatformToolset from project

snickler committed:
[ShortcutGuide] Removed PlatformToolset from project

snickler committed:
[ZoomIt] Removed PlatformToolset from project

snickler committed:
[keyboardmanager] Removed PlatformToolset from project

snickler committed:
[fancyzones] Removed PlatformToolset from project

snickler committed:
[imageresizer] Removed PlatformToolset from project

snickler committed:
[cmdpal] Removed PlatformToolset from project

snickler committed:
[launcher] Removed PlatformToolset from project

snickler committed:
[colorPicker] Removed PlatformToolset from project

snickler committed:
[powerrename] Removed PlatformToolset from project

snickler committed:
[runner] Removed PlatformToolset from project

snickler committed:
[registrypreview] Removed PlatformToolset from project

snickler committed:
[poweraccent] Removed PlatformToolset from project

snickler committed:
[peek] Removed PlatformToolset from project

snickler committed:
[previewpane] Removed PlatformToolset from project

snickler committed:
[tools] Removed PlatformToolset from project

snickler committed:
[Display] Add CppWinRT NuGet package and remove PlatformToolset

snickler committed:
[Common] Update async patterns to use C++/WinRT types

snickler committed:
[Workspaces] Update OnThreadExecutor usage

snickler committed:
[ZoomIt] Refactor CaptureScreenshotAsync to be synchronous

snickler committed:
[FancyZones] Update OnThreadExecutor usage

vanzue committed:
import defined props

vanzue committed:
Use global precompile header naming practice

vanzue committed:
fix warning

yeelam-gordon commented on src/common/updating/updating.cpp:
```diff
@@ -88,70 +88,78 @@ namespace updating
-                const auto body = co_await client.request(Uri{ LATEST_RELEASE_ENDPOINT });
-                release_object = json::JsonValue::Parse(body).GetObjectW();
-                if (auto extracted_version = extract_version_from_release_object(release_object))
+                else
```
The HttpClient.request() now returns IAsyncOperation&lt;hstring&gt; instead of std::future&lt;std::wstring&gt;, but the code calls .get() on it directly. IAsyncOperation doesn't have a .get() method - it needs to be awaited with co_await or you need to call .get() on the underlying winrt::impl interface. → This code likely doesn't compile. Either: (1) Keep coroutines and use co_await, or (2) Convert to synchronous blocking calls using winrt::impl::get_abi or similar blocking mechanism, or (3) Use proper async/await throughout the call chain.

yeelam-gordon commented on src/common/updating/updating.cpp:
```diff
@@ -172,30 +180,32 @@ namespace updating
-        for (size_t i = 0; i < MAX_DOWNLOAD_ATTEMPTS; ++i)
-        {
-            try
+        return std::async(std::launch::async, [new_version]() -> std::optional<std::filesystem::path> {
```
Similar issue as line 133: `client.download(...).get()` is called on IAsyncAction, which doesn't have a .get() method. The refactoring is incomplete. → Needs the same fix as the request() calls - either use co_await throughout or implement proper blocking mechanisms for IAsyncAction.

yeelam-gordon commented on src/common/updating/updating.cpp:
```diff
@@ -88,70 +88,78 @@ namespace updating
+                                                     std::move(installer_filename) };
+            }
+            catch (...)
             {
```
The catch block catches all exceptions (`catch (...)`) without any logging or error handling. In the original coroutine version, exceptions would propagate as part of the coroutine exception handling. Now they're silently swallowed and converted to NETWORK_ERROR. This could hide security-relevant errors (e.g., certificate validation failures, signature verification issues) and makes debugging harder. → Add logging for caught exceptions before converting to NETWORK_ERROR. Use Logger::error() to record the exception type and any available diagnostic information.

yeelam-gordon replied:
PowerToys uses centralized logging via `Logger::info/warn/error` from common/logger. The catch-all block silently swallows exceptions without logging. This violates the repo pattern of comprehensive logging for diagnostics. → Add `Logger::error("Update check failed", ...)` in the catch block to maintain consistency with other error handling in PowerToys.

yeelam-gordon commented on src/common/utils/OnThreadExecutor.h:
```diff
@@ -28,13 +29,17 @@ class OnThreadExecutor final
+        }
+
+        co_await winrt::resume_background();
+        future.wait();
```
The submit() method now returns IAsyncAction and uses co_await, but the function queues the task and then does `co_await winrt::resume_background(); future.wait();`. This creates a confusing execution model: (1) The task is queued on the executor thread, (2) The calling coroutine suspends and resumes on a background thread, (3) Then blocks that background thread waiting for the task to complete. → This defeats the purpose of both the executor pattern and async/await. The original std::future&lt;void&gt; return was more appropriate - callers could choose when/how to wait. If you need IAsyncAction for compatibility, consider not blocking internally and letting the IAsyncAction complete when the task finishes (use task completion callbacks).

yeelam-gordon commented on src/common/updating/updating.cpp:
```diff
@@ -88,70 +88,78 @@ namespace updating
+            std::expected<github_version_info, std::wstring>
 #else
-            co_return nonstd::make_unexpected(LOCAL_BUILD_ERROR);
+            nonstd::expected<github_version_info, std::wstring>
```
The refactoring from coroutines to std::async is a significant architectural change with no explanatory comment. Why was this necessary? Is it a VS 2026 coroutine bug workaround? A temporary measure? This needs documentation. → Add a comment like: '// Refactored from coroutines to std::async due to [reason]. See issue #[number] for details.' or '// VS 2026 Preview: Temporarily using std::async until C++20 coroutine support stabilizes.'

yeelam-gordon commented on src/common/utils/HttpClient.h:
```diff
@@ -21,15 +21,15 @@ namespace http
             auto body = co_await response.Content().ReadAsStringAsync();
-            co_return std::wstring(body);
+            co_return body;
         }
```
HttpClient methods changed return types from std::future&lt;T&gt; to IAsyncOperation&lt;T&gt;/IAsyncAction. This is a breaking API change for all callers. While the change aligns with WinRT patterns, all callsites must be updated. The updating.cpp file calls .get() on these, which suggests incomplete migration. → Ensure all consumers of HttpClient are updated: (1) Search for all HttpClient usage across the codebase, (2) Verify each callsite is updated to handle IAsync* types correctly, (3) Update any unit tests that mock or test HttpClient.

yeelam-gordon commented on src/modules/fancyzones/FancyZonesLib/on_thread_executor.h:
```diff
@@ -17,7 +18,7 @@ class OnThreadExecutor final
 
     OnThreadExecutor();
     ~OnThreadExecutor();
```
Code duplication: OnThreadExecutor exists in both common/utils and fancyzones/FancyZonesLib with nearly identical implementations. This violates DRY (Don't Repeat Yourself) and creates maintenance burden. Changes must be applied to both locations. → Consolidate to a single implementation in common/utils and remove the FancyZones duplicate, or clearly document why duplication is necessary (e.g., different threading models).

yeelam-gordon commented on Cpp.Build.props:
```diff
@@ -58,9 +53,15 @@
       <BuildStlModules>false</BuildStlModules>
-      <AdditionalOptions>/await %(AdditionalOptions)</AdditionalOptions>
+      <!-- Use C++20 for test projects for modern features, latest for production -->
+      <LanguageStandard Condition="$(MSBuildProjectName.ToLower().Contains('test'))">stdcpp20</LanguageStandard>
```
Comment says 'Use C++20 for test projects for modern features, latest for production' but doesn't explain why this split is necessary. This is a significant decision that future maintainers need to understand. → Expand comment to explain the reasoning: 'Use C++20 for test projects to ensure compatibility with [test framework name] while production uses stdcpplatest to leverage [specific features]' or document the ABI compatibility concerns.

yeelam-gordon reviewed (changes requested):
## VS 2026 Support Review - Critical Issues Found

I've completed a comprehensive review of this PR upgrading PowerToys from Visual Studio 2022 (v143) to VS 2026 (v145). While the intent is good, there are several **critical issues that will prevent compilation** and others that need attention.

### 🔴 Critical (Must Fix Before Merge)

1. **Incomplete API refactoring**: The HttpClient and OnThreadExecutor now return `IAsyncAction`/`IAsyncOperation` but the code calls `.get()` on them, which **doesn't exist on WinRT async types**. This likely won't compile.

2. **PlatformToolset fallback broken**: Checks for `VisualStudioVersion == '17.0'` but VS 2022 uses 17.x (17.1-17.11+). Should use `< '18.0'` instead.

3. **Coroutine ABI mismatch**: Test projects excluded from `_COROUTINE_ABI=2` and `/await:strict`, creating ABI incompatibility with production code.

### ⚠️ High Priority Issues

- **Performance regression**: Refactoring from coroutines to `std::async` with blocking `.get()` calls defeats async I/O benefits and risks thread pool starvation
- **Missing error logging**: Catch blocks swallow exceptions without logging, hiding security-relevant errors
- **Code duplication**: OnThreadExecutor exists in both `common/utils` and `fancyzones/FancyZonesLib`

### 📋 Documentation Updates Needed

Multiple docs reference "VS 2022" and need updating:
- BUILD-GUIDELINES.md
- doc/devdocs/modules/fancyzones.md  
- doc/devdocs/development/*.md
- README.md prerequisites
- CI/CD workflows

### Recommendations

1. **Fix the .get() calls**: Either revert to coroutines or use proper WinRT blocking mechanisms
2. **Correct the version check**: Use `'$(VisualStudioVersion)' < '18.0'`
3. **Add explanatory comments**: Document why the coroutine refactoring was necessary
4. **Consolidate duplicates**: Remove OnThreadExecutor duplication
5. **Verify CI/CD**: Ensure GitHub Actions runners have VS 2026 or fallback works

Full detailed review saved in review files. Happy to discuss any of these findings!

vanzue commented on src/modules/fancyzones/FancyZonesLib/EditorParameters.cpp:
```diff
@@ -136,7 +136,7 @@ bool EditorParameters::Save(const WorkAreaConfiguration& configuration, OnThread
             [&]() {
                 combinedWorkArea = FancyZonesUtils::GetAllMonitorsCombinedRect<&MONITORINFOEX::rcWork>();
-        } }).wait();
+        } }).get();
```
It builds well

snickler committed:
Revert "[Common] Update async patterns to use C++/WinRT types
"

This reverts commit 0d020e8321505e8dbd6948e6a9ed488fc3f1fd79.

snickler committed:
Revert "[Workspaces] Update OnThreadExecutor usage
"

This reverts commit 671993525d4f0c356cf83d7d8edd35e933702b38.

snickler committed:
Revert "[ZoomIt] Refactor CaptureScreenshotAsync to be synchronous
"

This reverts commit 37f90d8c1fcfb8d31352cdae4ec912e0c789ee0d.

snickler committed:
Revert "[FancyZones] Update OnThreadExecutor usage
"

This reverts commit dd76564a47696b7ef6f36b586474cafc19aef3c8.

snickler committed:
Revert "[build] Update PlatformToolset to v145 and configure C++ standard/coroutines
"

This reverts commit 9b037afed7c53f67603508d9ed9bff8fc70c00df.

snickler committed:
[build] Update PlatformToolset to 145 with fallback to 143

snickler committed:
[Workspaces] Undo /await:strict

snickler committed:
[Installer] Undo /await:strict

snickler committed:
[AlwaysOnTop] Undo /await:strict

snickler committed:
[FancyZones] Undo /await:strict

snickler committed:
[KeyboardManager] Undo /await:strict

snickler committed:
[Dev][CI] Updated to using VS18 compatible build agents

snickler committed:
[Build] Update vsVersion to 18.0 across build templates

snickler committed:
[Build] Update FuzzTest.props TargetFramework to net8.0-windows10.0.26100.0

snickler committed:
[Build] Add 4706 and 4874 warnings to be disabled.

snickler committed:
Merge remote-tracking branch 'origin/main' into dev/snickler/vs2026-support

snickler committed:
[CropAndLock]: Remove PlatformTarget

snickler committed:
[FileLockSmith]: Remove PlatformToolset from project

snickler committed:
[LightSwitch]: Remove PlatformToolset from project

snickler committed:
[CI][MouseJump][Tests] Adjust pixel color difference tolerance in tests

Increased tolerance for pixel color differences in tests to support minor variations in Windows Server 2025 CI.

snickler committed:
Remove trailing whitespace

snickler committed:
[DevDocs]: update documentation for Visual Studio 2026 support

snickler committed:
[Docs][Template] Updated README to reference VS 2026

snickler committed:
[Docs][Awake]: Update README to reference VS 2026

snickler committed:
[Docs][Build]: Update to reference VS 2026

snickler committed:
[Agents] Update AGENTS.md to include reference to VS 2026

yeelam-gordon committed:
[Build] Update VCTools version handling for VS2026 support

yeelam-gordon reviewed (approved):
Super sorry for my late

khmyznikov reviewed (approved):
LGTM
